### PR TITLE
Sample fine-tuning environment

### LINKED ISSUE #121: Illegal Memory Access with Mixed Precision
Hi, I'm running into a blocking issue when fine-tuning **Aurora** on a 95 GB H100 GPU using **Mixed Precision Training** (either via `bf16_mode` or `torch.autocast`).

---

### Setup

- **GPU:** NVIDIA H100 (95 GB)
- **Precision Mode:** BF16 (via `bf16_mode`) and/or `torch.autocast`
- **Batch Size:** 1 (with gradient accumulation over 4 steps)
- **Model:** Aurora
- **Fine-tuning Target:** 9 new **surface** variables  
- **Input to Model:** All pre-trained surface, static, and atmospheric variables  
- **Loss Function:** The Aurora loss as described in the original paper

---

###  What’s happening?

- The forward pass (`model(batch_in)`) runs **without any problems**.
- The error occurs **at `loss.backward()`**, when Mixed Precision is enabled.
- Specifically, I get an **illegal memory access** error that causes training to fail.

This only happens when Mixed Precision is active.

---

### What I’ve tried

- **Disabled BF16 & Autocast →**
  - ✅ No illegal memory access  
  - ❌ But eventually get an **out-of-memory (OOM)** error after 4–5 steps (full precision consumes about 90 GB during the first couple steps, and > 95 GB later)
  
- **Disabled Gradient Checkpointing (while keeping BF16 or Autocast enabled) →**
  - ✅ No illegal memory access  
  - ❌ But not viable because memory usage is too high without checkpointing

So it seems like the issue **only happens when Mixed Precision and Gradient Checkpointing are used together** ?
I could really use some help for debugging this.

### LINKED ISSUE #125: Multi-GPU training and mixed precision
Hi, thanks for this super nice work! 

I'm trying to train Aurora on multiple 95GB H100s but keep running into issues with mixed precision.

## Environment

- Python version: Python 3.11.7
- Pytorch version: 2.5.1+cu124
- Aurora version: 1.7.0

## Summary of the problem

I am unable to train Aurora (large) on more than 1 GPU. I have tried the suggestions in https://github.com/microsoft/aurora/issues/121 and https://github.com/microsoft/aurora/issues/119, namely using `autocast` rather than `bf16-mode` and changing the activation checkpointing but, regardless of the configuration, I eventually get either a Cuda OOM or an illegal memory access error.

I have actually created a repo to test these different configurations on our cluster using random data. It is publicly available [here](https://github.com/eliotwalt/aurora-testbed) and contains a much more complete description of the problem. Being relatively new to distributed training, I might be missing. Please find below a simple training script and the torchrun command I use to start it. 

## Training script

```python
import os
from tqdm import tqdm
import datetime
import contextlib
import argparse

import torch
import torch.nn as nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.utils.data.distributed import DistributedSampler
from torch.utils.data import DataLoader, TensorDataset
from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
    apply_activation_checkpointing,
)

from aurora.model.aurora import Aurora, AuroraSmallPretrained
from aurora import Batch, Metadata

torch.backends.cuda.matmul.allow_tf32 = True
torch.set_float32_matmul_precision("high")

# https://github.com/microsoft/aurora/issues/80#issuecomment-2849357112
os.environ["PYTORCH_CUDA_ALLOC_CONF"]="max_split_size_mb:32"

def make_batch(x, lat, lon):
    H, W = x.shape[-2:]
    return Batch(
        surf_vars={var: x[:,:,i] for i, var in enumerate(["10u", "10v", "2t", "msl"])},
        atmos_vars={var: x[:,:,i*13:(i+1)*13] for i, var in enumerate(["t", "u", "v", "q", "z"])},
        static_vars={var: torch.randn(H,W) for var in ["z", "lsm", "slt"]},
        metadata=Metadata(
            lat=lat, lon=lon, time=(datetime.datetime.now(),),
            atmos_levels=(50, 100, 150, 200, 250, 300, 400, 500, 600, 700, 850, 925, 1000),
        )
    )
    
def compute_loss(criterion, preds, target):
    loss = None
    for var_kind in ["surf_vars", "atmos_vars"]:
        for var_name in getattr(target, var_kind).keys():
            pred_var = getattr(preds, var_kind)[var_name]
            target_var = getattr(target, var_kind)[var_name]
            if loss is None: loss = criterion(pred_var, target_var)
            else: loss += criterion(pred_var, target_var)
    return loss

def custom_activation_checkpointing(model, module_names=("Perceiver3DEncoder","Swin3DTransformerBackbone","Basic3DEncoderLayer","Basic3DDecoderLayer","Perceiver3DDecoder","LinearPatchReconstruction")):
        def check(x: torch.nn.Module) -> bool:
            name = x.__class__.__name__
            return name in module_names
        apply_activation_checkpointing(model, check_fn=check)

def train(args):
    rank = int(os.environ["RANK"])
    local_rank = int(os.environ["LOCAL_RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
    if args.no_ddp and world_size > 1:
        raise ValueError("Distributed training is enabled but WORLD_SIZE > 1. Use --no_ddp to disable DDP.")
    dist.init_process_group(backend='nccl')
    device = local_rank # torch.device(f'cuda:{local_rank}')
    if rank == 0: print("Starting Aurora training script with arguments:", args)
    
    # Dummy dataset
    N = 5  # Number of samples
    M = 69 # 4 + 5 * 13
    H, W = 720, 1440
    lat=torch.linspace(90, -90, H)
    lon=torch.linspace(0, 360, W+1)[:-1]
    random_input = torch.randn(N, 2, M, H, W, dtype=torch.bfloat16)
    random_target = torch.randn(N, 1, M, H, W, dtype=torch.bfloat16)
    dataset = TensorDataset(random_input, random_target)
    sampler = DistributedSampler(dataset, shuffle=True)
    dataloader = DataLoader(dataset, sampler=sampler, batch_size=1, num_workers=1)
    if rank == 0: print("Dummy dataset created with random data.")
    
    # model
    if args.small: model = AuroraSmallPretrained(bf16_mode=args.bf16, use_lora=False)
    else: model = Aurora(bf16_mode=args.bf16, use_lora=False)
    model.load_checkpoint(strict=False)
    custom_activation_checkpointing(model, args.checkpointing_module_names)
    model = model.to(device)
    model = model.train()
    if not args.no_ddp:
        model = DDP(model, device_ids=[local_rank])
    if rank == 0: print("Model loaded and wrapped in DDP.")
    
    # loss and optimizer
    criterion = nn.L1Loss().to(device)
    optimizer = torch.optim.AdamW(model.parameters(), lr=5e-5, weight_decay=5e-6)
    if rank == 0: print("Loss and optimizer configured.")
    
    # autocast context
    if args.autocast: autocast_context = torch.autocast(device_type='cuda', dtype=torch.bfloat16)
    else: autocast_context = contextlib.nullcontext()
    
    # training loop
    global_step = 0
    for epoch in range(args.num_epochs):
        pbar = tqdm(dataloader, desc=f"Epoch {epoch+1}/{args.num_epochs}", disable=(local_rank != 0))
        for _, (inputs, targets) in enumerate(dataloader):
            # to batch and move to device
            inputs, targets = map(lambda x: make_batch(x, lat, lon)._fmap(lambda x: x.to(device, non_blocking=True)), (inputs, targets))
            optimizer.zero_grad()
            # fwd
            with autocast_context:
                outputs = model(inputs)
                loss = compute_loss(criterion, outputs, targets)
            # bwd 
            loss.backward()
            optimizer.step()
            # print
            pbar.update(1)
            pbar.set_postfix(loss=loss.detach().item(), step=global_step)
            global_step += 1
                
def main(args):
    try: train(args)
    except Exception as e: dist.destroy_process_group() ; raise e
    
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--num_epochs", type=int, default=10, help="Number of epochs to train for")
    parser.add_argument("--small", action="store_true", help="Use the small version of the Aurora model")
    parser.add_argument("--bf16", action="store_true", help="Use `bf16_mode`")
    parser.add_argument("--autocast", action="store_true", help="Use autocast context with bf16")
    parser.add_argument("--checkpointing_module_names", type=str, nargs='+', 
                        default=["Perceiver3DEncoder","Swin3DTransformerBackbone","Basic3DEncoderLayer","Basic3DDecoderLayer","Perceiver3DDecoder","LinearPatchReconstruction"])
    parser.add_argument("--no_ddp", action="store_true", help="Disable DDP")
    args = parser.parse_args()
    main(args)
```

## Launch command

```bash
 srun torchrun --standalone --nnodes=1 --nproc_per_node=2 \
            train.py --num_epochs=500 --bf16 \
                --checkpointing_module_names Perceiver3DEncoder Swin3DTransformerBackbone Basic3DEncoderLayer Basic3DDecoderLayer Perceiver3DDecoder LinearPatchReconstruction
```

Does that look reasonable to you? I would appreciate any recommendations you might have :) Thanks!

### PR DESCRIPTION
**Pull request by wesselb**
- [x] Removes `bf16_mode` and brings back `autocast` (Closes #121.).
- [x] Make activation checkpointing configurable and improve default checkpointing strategy.
- [x] Change default drop-out rates to zero everywhere
- [x] Fix bug: Set `attn_drop_rate` in the backbone. This will cause stochasticity if the model wasn't `.eval()`d.
- [x] Adds a Docker image which runs a very basic fine-tuning loop for Aurora. (Closes #125. Closes #121.)

I've verified that the instructions work on a fresh Azure VM.

### CODE REVIEW

File: docs/finetuning.md [hunk 1]
```diff
@@ -10,6 +10,44 @@ model = AuroraPretrained()
...
+## Basic Fine-Tuning Environment
+
+We provide a very basic Docker image and fine-tuning loop to get you started.
+The Docker image can be found at `finetuning/Dockerfile` and the fine-tuning
+loop at `finetuning/finetune.py`.
+You can build and run the image as follows:
+
+```bash
+docker build . -t aurora:latest -f finetuning/Dockerfile
+docker run --rm -it -v .:/app/aurora \
```
Comment (rchan26) [inline docs/finetuning.md]: An issue may arise here if the user already has a `.venv` folder locally as that would get mounted. I think to solve this, you need to put an anonymous mount like in https://docs.astral.sh/uv/guides/integration/docker/#mounting-the-project-with-docker-run

In particular, you might need to do `-v /app/.venv`

File: docs/finetuning.md [hunk 2]
```diff
@@ -10,6 +10,44 @@ model = AuroraPretrained()
...
+    --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 \
+    aurora:latest
+```
+
+Then, within the image, execute
+
+```bash
+python finetuning/finetune.py
```
[the remark this answers is not in the document]
Comment (rchan26) [reply docs/finetuning.md]: Otherwise, you could just turn on the environment by default in the docker container by adding `echo "source /app/.venv/bin/activate" > ~/.bashrc`

File: finetuning/finetune.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        metadata=Metadata(
+            lat=torch.linspace(90, -90, 721),
+            lon=torch.linspace(0, 360, 1440 + 1)[:-1],
+            time=(datetime(2020, 6, 1, 12, 0),),
+            atmos_levels=(50, 100, 150, 200, 250, 300, 400, 500, 600, 700, 850, 925, 1000),
+        ),
+    )
+
+    opt.zero_grad()
+    prediction = model.forward(batch.to("cuda"))
```
Comment (rchan26) [inline finetuning/finetune.py]: Is `.forward()` necessary?

```suggestion
    prediction = model(batch.to("cuda")
```

**Review by rchan26 — COMMENTED** [review]
Thanks @wesselb - this looks great! Just some small comments. The main one is about ensuring that we definitely use the right `.venv` when in the image. Once we've made sure that is the case, I think we can merge this in :)
